packages/tpetra/core/src/Tpetra_Distributor.hpp

```diff
@@ -261,7 +268,21 @@ namespace Tpetra {
     ///
     /// \pre No outstanding communication requests.
     ///   (We could check, but see GitHub Issue #1303.)
-    virtual ~Distributor () = default;
+    ~Distributor () {
+      for (int i = 0; i<persistentRequests_.size(); i++) {
+        Teuchos::RCP<Teuchos::CommRequest<int> > request = persistentRequests_[i];
+        Teuchos::RCP<Teuchos::MpiCommRequestBase<int> > mpiRequest =
+          Teuchos::rcp_dynamic_cast<Teuchos::MpiCommRequestBase<int> > (request);
+        MPI_Request mR = mpiRequest->releaseRawMpiRequest();
```

Comment:
Let's talk about better encapsulation strategies.  I suspect that you don't have to do anything here; CommRequest's virtual destructor will dispatch to MpiCommRequest's destructor, which will cancel and free the outstanding requests.